[리액트 7주차 김주빈] 비동기 과제 제출합니다.

KJoobin opened the pull request:
비동기 관련해서는 테스트를 어떻게 짜야할지 좀더 고민해봐야할거같습니다 ㅠㅠ

강의자료와 같이 services 를 mocking 한 다음에 calledTimes 만으로 테스트를 진행하면 되는걸까요??!

KJoobin committed:
카테고리 렌더 테스트

KJoobin committed:
렌더 카테고리 구현

KJoobin committed:
렌더 카테고리 서비스 분리 리팩토링

KJoobin committed:
redux store, reducer 로 리팩토링

KJoobin committed:
불필요한 html 태그 삭제

KJoobin committed:
카테고리 클릭 이벤트 테스트

KJoobin committed:
카테고리 클릭 구현

KJoobin committed:
카테고리 클릭 인자 테스트

KJoobin committed:
카테고리 클릭 이벤트 테스트

KJoobin committed:
selectCategory 구현

KJoobin committed:
selectCateogry action 테스트

KJoobin committed:
reducer selectCategoryId 추가

KJoobin committed:
reducer selectCategory 테스트

KJoobin committed:
selectCategory 리듀서 구현

KJoobin committed:
selectCategory 인자 타입 리팩토링

-from number to object

KJoobin committed:
선택된 카테고리 텍스트 테스트

KJoobin committed:
선택된 카테고리 이름 구현

KJoobin committed:
Container 에서 선택된 카테고리 테스트

KJoobin committed:
Container 에서 선택된 카테고리 구현

KJoobin committed:
categories fixture 생성

KJoobin committed:
주소목록 렌더링 테스트

KJoobin committed:
주소 목록 렌더링 구현

KJoobin committed:
주소 리스트 서비스 분리 리팩토링

KJoobin committed:
주소 선택 이벤트 테스트

KJoobin committed:
주소 선택 클릭 이벤트 구현

KJoobin committed:
주소 선택 액션 테스트

KJoobin committed:
select address action 구현

KJoobin committed:
selecte address reducer 테스트

KJoobin committed:
select address reducer 구현

KJoobin committed:
selected address render 테스트

KJoobin committed:
select address 구현

KJoobin committed:
오타 수정 리팩토링

KJoobin committed:
name functions 분리 리팩토링

KJoobin committed:
구현했던 nameFunctions 를 helper 함수로 replace 리팩토링

KJoobin committed:
address 대신 region 으로 이름 리팩토링

KJoobin committed:
redux-thunk 설치

KJoobin committed:
fix code style

moonkii commented on src/Categories.jsx:
```diff
@@ -0,0 +1,21 @@
+          {nameFunctions[category.id === selectCategoryId](category.name)}
+        </button>
+      ))}
+    </ul>
```
의미있는 마크업을 하는 것도 중요해요.
`ul` 아래에는 `li`가 오죠.

moonkii commented on src/Categories.test.jsx:
```diff
@@ -0,0 +1,39 @@
+import categories from '../fixture/categories';
+
+describe('Categories', () => {
+  const onSelect = jest.fn();
```
mock function 을 사용하시면 초기화를 해주시면 좋아요.

moonkii commented on src/CategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,61 @@
+    });
+  });
+
+  context('when selected', () => {
```
무엇이 selected 된건가요?

moonkii commented on src/CategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,61 @@
+          { id: 1, name: '한식' },
+        ],
+      }));
+
```
맥락에 따라 바뀌는 부분은 given2 라이브러리를 사용해주시면 도움이 되실거에요;)

moonkii commented on src/reducer.test.js:
```diff
@@ -0,0 +1,43 @@
+      selectRegionId: undefined,
+      regions: [],
+    });
+  });
```
예외적인 케이스인 경우 context 를 사용해주시면 좋아요.

moonkii reviewed (commented):
![image](https://user-images.githubusercontent.com/46776329/163828462-4a583d79-607c-4f49-a93e-377e1cf56104.png)

테스트 코드도 잘 작성하셨고, 잘 진행하고 있으신데
비슷한 구조가 반복되는 것 같기도해도 리팩터링해보시면 좋을 것 같아요!

> 비동기 관련해서는 테스트를 어떻게 짜야할지 좀더 고민해봐야할거같습니다 ㅠㅠ

비동기 함수가 action 으로 들어갈 것이고, 그것에 대해 테스트해보면 될거에요.

> 강의자료와 같이 services 를 mocking 한 다음에 calledTimes 만으로 테스트를 진행하면 되는걸까요??!

api mocking 을 의미하시는 걸까요?
`calledTimes` 보다 의미가 더 드러나는 테스트 함수가 있을거에요.

KJoobin committed:
api services 추가

KJoobin committed:
fetch regions, categories 테스트

KJoobin committed:
setRegions 구현

KJoobin committed:
fetchRegions 테스트

KJoobin committed:
load Regions 구현

KJoobin committed:
setCategories 테스트

KJoobin committed:
setCategories 구현

KJoobin committed:
loadCategories 테스트

KJoobin committed:
loadCategories 테스트 구현

KJoobin committed:
categories, regions 를 selectList 로 리팩토링

KJoobin committed:
api 반복되는부분 리팩토링

KJoobin committed:
api 테스트

KJoobin committed:
feetchRestaurants api 테스트

KJoobin committed:
fetch restaurants 구현

KJoobin committed:
setRestaurants reducer 테스트

KJoobin committed:
setRestaurants reducer 구현

KJoobin committed:
loadRestaurants reducer 테스트

KJoobin committed:
loadRestaurants 구현

KJoobin committed:
load restaurants regionName 이 없을경우 테스트

KJoobin committed:
loadRestaurants regionName 이 없을경우 구현

KJoobin committed:
categoryId arg 가 없을경우 테스트

KJoobin committed:
categoryId 가 없을경우 구현

KJoobin committed:
service api mock 수정

- regionName, categoryId 가 없을경우에는 빈 배열 반환

KJoobin committed:
categoryId, regionName 모두 넘겼을때 테스트

KJoobin committed:
redux-thunk mocking 주석

KJoobin committed:
context 설명문 수정

KJoobin committed:
맥락에 따라 바뀌는 부분 given2 로 리팩토링

moonkii commented on src/CategoriesContainer.jsx:
```diff
@@ -0,0 +1,28 @@
+  }
+
+  return (
+    <SelectList
```
SelectList 이라는 이름은 직관적이진 않은 것 같아요.
어떻게 바꿔보면 좋을까요?

KJoobin replied:
카테고리, 지역을 보여주는 컴포넌트가 중복되는거 같아서 한개로 통합해서 재사용하기위해서 만들었는데, 두 의미를 모두 통합하는 직관적인 이름이 고민을 해도 잘 모르겠습니다... ㅠㅠ.....

노아님께서는 컴포넌트의 이름을 지을때 어떤점을 가장 고민하시고, 어떤걸 참고하시나요..?

moonkii replied:
공통 컴포넌트라면 list 정도로 해서 사용해주셔도 괜찮아요.

moonkii replied:
저는 컴포넌트 이름을 지을 때 먼저 범용인지, 도메인과 관련된 건지 고민을 하고 도메인을 드려내려고 하면 의미를 잘 드러나도록 고민을 하는편이에요.

moonkii commented on src/reducer.test.js:
```diff
@@ -0,0 +1,160 @@
+  describe('loadRestaurants', () => {
+    const dispatch = jest.fn();
+
+    context("when doesn't pass regionName arg", () => {
+      it("doesn't working", async () => {
+        await loadRestaurants({ categoryId: 1 })(dispatch);
+
+        expect(dispatch).not.toBeCalled();
+      });
+    });
+
+    context("when doesn't pass categoryId arg", () => {
+      it("doesn't working", async () => {
+        await loadRestaurants({ regionName: '서울' })(dispatch);
+
+        expect(dispatch).not.toBeCalled();
+      });
+    });
+
+    context('when pass categoryId and regionName', () => {
+      it('called setRestaurants with restaurants data', async () => {
```
맥락에 대한 설명을 좀 더 풀어서 해주시면 좋을 것 같아요.
지금 테스트 시나리오를 읽어서는 어떤 건지 이해하기 어려운 것 같아요.

moonkii commented on src/services/api.test.js:
```diff
@@ -0,0 +1,35 @@
+    jest.clearAllMocks();
+  });
+
+  const fetch = jest.fn(() => Promise.resolve({
+    json: () => Promise.resolve(''),
+  }));
+
+  global.fetch = fetch;
+
```
mocking 을 잘 하셨네요 👍

KJoobin committed:
context 설명문 변경

KJoobin committed:
redux-thunk 에서 state 얻는방법 리팩토링 테스트

KJoobin committed:
redux thunk 에서 state 값을 가져오는 방법 리팩토링

KJoobin committed:
typo 수정 테스트

KJoobin committed:
typo 수정

KJoobin committed:
RestaurantsContainer 테스트

KJoobin committed:
Restaurants 테스트

KJoobin committed:
Restaurants 구현

KJoobin committed:
RestaurantsContainer 구현

KJoobin committed:
App 에 RestaurantsContainer 가져오기

KJoobin committed:
헬퍼 함수 지우고, select list 안으로 move

KJoobin commented:
redux-thunk 를 테스트하기위해 다른방법을 사용해봤는데, 맞을까요...? ㅠ 너무 아닌거같네용..

moonkii commented on src/reducer.test.js:
```diff
@@ -128,23 +128,23 @@ describe('Reducer', () => {
     const dispatch = jest.fn();
 
-    context("when doesn't pass regionName arg", () => {
+    context('before select category', () => {
```
context 에서는 주로 with, without, when 을 많이 사용해요

참고: https://www.betterspecs.org/#contexts

moonkii commented on src/Restaurants.jsx:
```diff
@@ -0,0 +1,15 @@
+export default function Restaurants({ restaurants }) {
+  if (!restaurants) {
+    return null;
```
`null` 보다는 예외처리 메시지를 넣어주면 좋을 것 같아요!

moonkii commented on src/reducer.test.js:
```diff
@@ -0,0 +1,207 @@
+        categoryId: 1,
+      });
+
+      it("doesn't working", async () => {
```
working 한다 안한다는 건 어떤것도 드러내지 못하는 것 같아요.
더 의미를 드러내는 테스트 시나리오를 작성해볼 수 있을까요?

moonkii commented on src/services/api.test.js:
```diff
@@ -0,0 +1,35 @@
+
+  global.fetch = fetch;
+
+  it('called with regions url', async () => {
```
과거형을 잘 사용하지 않아요. calls 같이 현재형을 작성해주는게 더 의미가 맞을 것 같아요.

KJoobin committed:
테스트 설명 수정

- context 는 when, with, without을 주로 사용
- 과거형 보다는 현재형을 사용
- doesn't working 은 아무것도 들어내지 못하므로 직관적으로 수정

KJoobin committed:
테스트 설명문 수정

- 과거형 보다는 현재형을 사용

KJoobin committed:
SelectList 이름 List 로 리팩토링

KJoobin committed:
region 을 선택하기 전 테스트

KJoobin committed:
레스토랑 without select region 테스트

KJoobin committed:
지역을 선택하지 않았을때 메세지 구현

KJoobin committed:
지역을 선택하지 않았을때, restaurantsContainer 구현

KJoobin committed:
cateogry 를 선택하지 않았을때 메세지 테스트

KJoobin committed:
분류를 선택하지 않았을때 구현

KJoobin committed:
RestaurantsContainer 분류를 선택하지 않았을떄 테스트

KJoobin committed:
RestaurantsContainer 분류를 선택하지 않았을때 구현

KJoobin committed:
load restaurants 조건이 맞지않으면 call setRestaurants with empty array 리팩토링

KJoobin committed:
reducer loading state 테스트

KJoobin committed:
setLoading action 테스트

KJoobin committed:
setLoading, loading state 구현

KJoobin committed:
레스토랑 로딩 테스트

KJoobin committed:
로딩상태일때 메시지 구현

KJoobin committed:
레스토랑이 목록이 비어있을경우 구현

KJoobin committed:
RestaurantsContainer 테스트

KJoobin committed:
로딩 상태일때 구현

moonkii commented on src/reducer.test.js:
```diff
@@ -0,0 +1,245 @@
+    jest.clearAllMocks();
+  });
+
+  context('when call reducer with undefined args', () => {
```
with 정도로만 표현해줘도 괜찮을 것 같아요.

KJoobin committed:
테스트 코드 컨택스트 수정

KJoobin committed:
reducer 테스트 컨택스트 수정

moonkii reviewed (approved):
테스트 작성하기가 쉽지 않으셨을텐데 잘 작성하신 것 같아요 👍 
주간회고 작성 후 공유해주시면 merge 하도록 하겠습니다!
